torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+        TORCH_CHECK(output_index < static_cast<int>(state.outputs.size()));
+        TORCH_CHECK(!state.outputs[output_index].defined());
+        state.outputs[output_index] = inputs[input_nr];
+      }
+
+      if (call.tensor_pre_hooks.size() + call.pre_hooks.size() > 0) {
+        // TODO(jansel): we should lift hooks to be inputs to the graph since we
+        // are not specializing on them
+        THPObjectPtr pyinputs(wrap_variable_list(inputs));
+        for (const auto& hook : call.tensor_pre_hooks) {
+          pyinputs = check(PyObject_CallMethod(
+              py_compiler,
+              "tensor_pre_hook",
+              "Oii",
+              pyinputs.get(),
+              hook.first,
+              hook.second));
+        }
+        for (const auto hook : call.pre_hooks) {
+          pyinputs = check(PyObject_CallMethod(
+              py_compiler.get(), "pre_hook", "Oi", pyinputs.get(), hook));
+        }
+        inputs = unwrap_variable_list(pyinputs);
+      }
+
+      SwapSavedVariables saved(state, call.node);
+      variable_list outputs = call.node->apply_with_saved(inputs, saved);
+
+      if (call.post_hooks.size() > 0) {
+        THPObjectPtr pyinputs(wrap_variable_list(inputs));
+        THPObjectPtr pyoutputs(wrap_variable_list(outputs));
+        for (const auto hook : call.post_hooks) {
+          pyoutputs = check(PyObject_CallMethod(
+              py_compiler.get(),
+              "post_hook",
+              "OOi",
+              pyoutputs.get(),
+              pyinputs.get(),
+              hook));
+        }
+        outputs = unwrap_variable_list(pyoutputs);
+      }
+
+      node_outputs.emplace_back(std::move(outputs));
+    }
+
+    cache->compiled_fn = check(call_end_capture(py_compiler, state.outputs));
+  }
+
+  // TODO(jansel): we should release all the variables and then use a
+  //               boxed calling convention so activation memory can be freed
+  for (auto& call : calls) {
+    call.node->release_variables();
+  }
+
+  {
+    THPObjectPtr inputs(wrap_variable_list(compiler_call.inputs));
+    THPObjectPtr sizes(wrap_int_list(compiler_call.dyn_size_inputs));
+    THPObjectPtr hooks(convert_hook_list(compiler_call.hooks));
+    THPObjectPtr pyresult(check(PyObject_CallFunctionObjArgs(
+        cache->compiled_fn.get(),
+        inputs.get(),
+        sizes.get(),
+        hooks.get(),
+        NULL)));
+    variable_list outputs = unwrap_variable_list(pyresult);
+    if (accumulate_grad) {
+      TORCH_CHECK(outputs.size() == compiler_call.set_grad_targets.size());
+      for (const auto i : c10::irange(outputs.size())) {
+        // TODO(jansel): does this one need to be an inplace copy?  if so it
+        // should go in the graph
+        at::Tensor& grad = compiler_call.set_grad_targets[i].mutable_grad();
+        grad = outputs[i];
+      }
+      return variable_list();
+    } else {
+      TORCH_CHECK(outputs.size() == output_edges.size());
+      return outputs;
+    }
+  }
+}
```

Comment:
I'm going to let the autograd peeps do the careful review on the main loop of algo